File: Python/candy.py
class Solution:
    # @param ratings, a list of integer
    # @return an integer
    def candy(self, ratings):
        
        n = len(ratings)
        
        candy = [1]*n
        
        for i in xrange(1, n):
            if ratings[i]>ratings[i-1]:
                candy[i] = candy[i-1]+1

        for i in xrange(n-1, 0, -1):
            if ratings[i-1]>ratings[i] and candy[i-1]<=candy[i]:
                candy[i-1] = candy[i]+1

        return sum(candy)

File: C++/candy.cpp
class Solution {
public:
    int candy(std::vector<int> &ratings) {
        int n = ratings.size();
        std::vector<int> candies (n, 1);
        for(int i=1; i<n; ++i)
            if(ratings[i]>ratings[i-1])
                candies[i] = 1+candies[i-1];
        for(int i= n-1; i>0; --i)
            if(ratings[i-1]>ratings[i] && candies[i-1]<=candies[i])
                candies[i-1] = 1+candies[i];
        return std::accumulate(candies.begin(), candies.end(), 0);
    }
};
